element-connector/src/main/java/org/eclipse/californium/elements/tcp/TcpServerConnector.java

```diff
@@ -82,17 +85,17 @@ public TcpServerConnector(InetSocketAddress localAddress, int numberOfThreads, i
 				.option(ChannelOption.AUTO_READ, true).childHandler(new ChannelRegistry());
 
 		// Start the server.
-		bootstrap.bind(localAddress).syncUninterruptibly().addListener(new ChannelFutureListener() {
-
-			@Override
-			public void operationComplete(ChannelFuture future) throws Exception {
-				InetSocketAddress listenAddress = (InetSocketAddress) future.channel().localAddress();
-				listenUri = getListenUri(listenAddress);
-			}
-		});
+		ChannelFuture channelFuture = bootstrap.bind(localAddress).syncUninterruptibly();
+
+		if (channelFuture.isSuccess() && 0 == localAddress.getPort()) {
+			InetSocketAddress listenAddress = (InetSocketAddress) channelFuture.channel().localAddress();
+			localAddress = new InetSocketAddress(localAddress.getAddress(), listenAddress.getPort());
```

Comment:
Oh, now I see that you are using `localAddress.getAddress()` instead of `listenAddress.getAddress()`.

The problem lies in the `getListenUri()` method. We need to distinguish between `Inet4Address` and `Inet6Address` because the URI constructor requires IPv6 literal addresses to be enclosed in square brackets:

```
private URI getListenUri(final InetSocketAddress listenAddress) {

  InetAddress address = listenAddress.getAddress();
  String addressString = address.getHostString();
  if (address instanceof Inet6Address) {
    addressString = String.format("[%s]", addressString);
  }
      return URI.create(String.format("%s://%s:%d", getSupportedScheme(), addressString, address.getPort()));
}
```